README.md

```diff
@@ -36,80 +38,152 @@ bundle install
+| `key`       | false    | String        | The key that should be set in the Kafka message       |
+| `partition` | false    | Integer       | A specific partition number that should be written to |
+| `timestamp` | false    | Time, Integer | The timestamp that should be set on the message       |
+| `headers`   | false    | Hash          | Headers for the message                               |
+
+Keep in mind, that message you want to send should be either binary or stringified (to_s, to_json, etc).
 
-To send Kafka messages, just use one of the producers:
+### Buffering
+
+WaterDrop producers support buffering of messages, which means that you can easily implement periodic flushing for long running processes as well as buffer several messages to be flushed the same moment:
 
 ```ruby
-WaterDrop::SyncProducer.call('message', topic: 'my-topic')
-# or for async
-WaterDrop::AsyncProducer.call('message', topic: 'my-topic')
+producer = WaterDrop::Producer.new
+
+producer.setup do |config|
+  config.kafka = { 'bootstrap.servers' => 'localhost:9092' }
+end
+
+time = Time.now - 10
+
+while time < Time.now
+  time += 1
+  producer.buffer(topic: 'times', payload: Time.now.to_s)
+end
+
+puts "The messages buffer size #{producer.messages.size}"
+producer.flush_sync
+puts "The messages buffer size #{producer.message.size}"
+
+producer.close
 ```
 
-Both ```SyncProducer``` and ```AsyncProducer``` accept following options:
+## Instrumentation
 
-| Option              | Required | Value type | Description                                                         |
-|-------------------- |----------|------------|---------------------------------------------------------------------|
-| ```topic```         | true     | String     | The Kafka topic that should be written to                           |
-| ```key```           | false    | String     | The key that should be set in the Kafka message                     |
-| ```partition```     | false    | Integer    | A specific partition number that should be written to               |
-| ```partition_key``` | false    | String     | A string that can be used to deterministically select the partition |
-| ```create_time```   | false    | Time       | The timestamp that should be set on the message                     |
-| ```headers```       | false    | Hash       | Headers for the message                                             |
+Each of the producers after the `#setup` is done, has a custom monitor to which you can subscribe.
 
-Keep in mind, that message you want to send should be either binary or stringified (to_s, to_json, etc).
+```ruby
+producer = WaterDrop::Producer.new
+
+producer.setup do |config|
+  config.kafka = { 'bootstrap.servers' => 'localhost:9092' }
+end
+
+producer.monitor.subscribe('message.produced_async') do |event|
+  puts "A message was produced to '#{event[:message][:topic]}' topic!"
+end
+
+producer.produce_async(topic: 'events', payload: 'data')
+```
```

Comment:
Maybe it would be better to add in the end the line`producer.close` to have the documentation consistent.